### PR TITLE
feat(docker/artifactory): add retry logic for docker offical images

### PR DESCRIPTION
**Pull request by setchy**
## Changes

Introduce an additional retry condition for retrieving docker tags when using JFrog.IO Artifactory [repositories](https://www.jfrog.com/confluence/display/JFROG/Repository+Management) and [official images](https://docs.docker.com/docker-hub/official_images/)

This new condition is an adaption of the [existing docker tag retry logic](https://github.com/renovatebot/renovate/blob/af75ec25084d67c6239302ccf6dd1cbe1eec40ef/lib/modules/datasource/docker/index.ts#L891).

## Context

JFrog.io Artifactory has the concept of [repositories](https://www.jfrog.com/confluence/display/JFROG/Repository+Management) (local, remote, virtual,  etc), which are referenced in docker commands such as follows:

```
docker pull <ORG_NAME>.jfrog.io/<REPOSITORY_NAME>/<DOCKER_IMAGE>:<DOCKER_TAG>
```

Some working examples include:

- `docker pull org.jfrog.io/docker/node:latest`
- `docker pull org.jfrog.io/docker/dotnet/sdk:latest`

And the corresponding docker tags/list REST API calls should be:

- `https://org.jfrog.io/v2/docker/library/node/tags/list`
- `https://org.jfrog.io/v2/docker/dotnet/aspnet/tags/list`

However, Renovate's current behavior for fetching Official Images (eg: `node`) when combined with JFrog.io results in these two REST API calls, which both return `HTTP 404 Not Found`

- `https://org.jfrog.io/v2/docker/node/tags/list` 
- `https://org.jfrog.io/v2/library/docker/node/tags/list`

This PR adds new conditional behavior to inject `library/` into the REST API path between the `<REPOSITORY_NAME>` and `<DOCKER_IMAGE>` 

Reference: [JFrog Artifactory REST API - List Docker Tags](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-ListDockerTags)

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Current behavior
renovate-current-behavior

Tested behavior on demo-project
renovate-new-tested-behavior

### CODE REVIEW

File: lib/modules/datasource/docker/index.ts [hunk 1]
```diff
@@ -897,6 +897,26 @@ export class DockerDatasource extends Datasource {
         );
         return this.getTags(registryHost, 'library/' + dockerRepository);
       }
+      // Retry handling for JFrog Artifactory when resolving Docker Official Images
+      // These follow the format of {{registryHost}}{{jFrogRepository}}/library/{{dockerRepository}}
+      if (
+        (err.statusCode === 404 || err.message === PAGE_NOT_FOUND_ERROR) &&
+        dockerRepository.split('/').length === 2 &&
+        registryHost.includes('jfrog.io')
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:905]: isn't this better detectable via server response headers?

Comment (setchy) [reply lib/modules/datasource/docker/index.ts:905]: Certainly - in fact that would then also support anyone who runs a custom domain name for their jfrog.io instance 😄 

There are 3 headers we would have access to

- `x-artifactory-id`
- `x-artifactory-node-id`
- `x-jfrog-version`

Any preference?

Comment (setchy) [reply lib/modules/datasource/docker/index.ts:905]: Decided to use `x-jfrog-version`.  The two others look to be correlation ids

**Comment by chezsmithy** [comment]
There is a further problem with jfrog. If you are using a virtual docker repo, the tags api calls will only return what's been pulled vs what is available upstream. Makes renovate fail to find new versions or digests.

**Comment by setchy** [comment]
> There is a further problem with jfrog. If you are using a virtual docker repo, the tags api calls will only return what's been pulled vs what is available upstream. Makes renovate fail to find new versions or digests.

😞 - yes, that too is another problem

**Comment by setchy** [comment]
@chezsmithy - I noticed in their [List Docker Tags api documentation](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerTags) this callout

> Notes: To enable fetch from cache using the [ListDockerRepositories](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerRepositories) and the [ListDockerTags](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerTags) rest APIs, set the artifactory.docker.catalogs.tags.fallback.fetch.remote.cache system property to true (default false) in the artifactory.system.properties file.

Have you got any hands-on experience with what this flag does (or doesn't do)?

File: lib/modules/datasource/docker/index.ts [hunk 2]
```diff
@@ -897,6 +897,26 @@ export class DockerDatasource extends Datasource {
         );
         return this.getTags(registryHost, 'library/' + dockerRepository);
       }
+      // JFrog Artifactory - Retry handling when resolving Docker Official Images
+      // These follow the format of {{registryHost}}{{jFrogRepository}}/library/{{dockerRepository}}
+      if (
+        (err.statusCode === 404 || err.message === PAGE_NOT_FOUND_ERROR) &&
+        err.headers['x-jfrog-version'] !== null &&
```
Comment (rarkins) [inline lib/modules/datasource/docker/index.ts:904]:
```suggestion
        is.string(err.headers['x-jfrog-version']) &&
```

File: lib/modules/datasource/docker/index.ts [hunk 3]
```diff
@@ -897,6 +897,26 @@ export class DockerDatasource extends Datasource {
...
+        dockerRepository.split('/').length === 2
```
Comment (rarkins) [inline lib/modules/datasource/docker/index.ts:905]: What's an example here? e.g. do you mean `some.artifactory.server/some-library-image-name`?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/docker/index.ts (hunk 3 above)
Comment (setchy) [reply lib/modules/datasource/docker/index.ts:905]: The `/library/` path is only required for [Docker Official Images](https://hub.docker.com/search?q=&image_filter=official).

Examples of these when going direct to Docker Hub include:

- `node`
- `python`
- `redis`
...

When used with JFrog Artifactory virtual repositories, these become:

- `some.artifactory.server/some.artifactory.repository/node`
- `some.artifactory.server/some.artifactory.repository/python`
- `some.artifactory.server/some.artifactory.repository/redis`
- ...

Hence checking for the presence of 2 `/` in the name.

Comment (rarkins) [reply lib/modules/datasource/docker/index.ts:905]: To clarify, `dockerRepository` in this case would be `some.artifactory.repository/node` and not the full string as shown above? And there's always a repository in the path (never the root) and the repository will never have a `/` within it?

Comment (setchy) [reply lib/modules/datasource/docker/index.ts:905]: > To clarify, dockerRepository in this case would be some.artifactory.repository/node and not the full string as shown above?

`dockerRepository` on line 905 contains the full string extracted from a Dockerfile FROM statement, so in this case it'd be `some.artifactory.server/some.artifactory.repository/node`

> And there's always a repository in the path (never the root) and the repository will never have a / within it?

That's correct - there must always be be `some.artifactory.repository` used within the Dockerfile FROM statement.

Here is a screengrab of the jfrog.io `set me up` screen for Docker
Screen Shot 2022-09-25 at 06 02 16

`some.artifactory.repository` will never have a `/` in it.

The `image` however might, but for those instances the current Renovate logic works perfectly with the JFrog Docker Tags List API

File: lib/modules/datasource/docker/index.ts [hunk 4]
```diff
@@ -901,7 +901,7 @@ export class DockerDatasource extends Datasource {
       // These follow the format of {{registryHost}}{{jFrogRepository}}/library/{{dockerRepository}}
       if (
         (err.statusCode === 404 || err.message === PAGE_NOT_FOUND_ERROR) &&
-        is.string(err.headers['x-jfrog-version']) &&
+        isArtifactoryServer(err) &&
```
Comment (rarkins) [inline lib/modules/datasource/docker/index.ts:904]: In this case is `err` compatible with `res: HttpResponse` ?

Comment (viceice) [reply lib/modules/datasource/docker/index.ts:904]: no. needs to be err.response or something like that

Comment (setchy) [reply lib/modules/datasource/docker/index.ts]: it was compatible in this case, however changed it to `err.response` to be clearer

**Comment by setchy** [comment]
@rarkins @viceice - any further feedback re this PR that you'd like me to address?

**Comment by setchy** [comment]
> There is a further problem with jfrog. If you are using a virtual docker repo, the tags api calls will only return what's been pulled vs what is available upstream. Makes renovate fail to find new versions or digests.

@chezsmithy - I got some good news from JFrog Support about your comment.  They have already fixed this issue in https://www.jfrog.com/jira/browse/RTFACT-27278, with a public release slated for Q4-2022.  In the interim, they suggested to use the remote docker repository directly, instead of a virtual docker repository. That solution isn't too practical, so we're holding off until their public release of the fix.

**Comment by chezsmithy** [comment]
> @chezsmithy - I noticed in their [List Docker Tags api documentation](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerTags) this callout
> 
> 
> 
> > Notes: To enable fetch from cache using the [ListDockerRepositories](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerRepositories) and the [ListDockerTags](https://www.jfrog.com/confluence/display/JCR6X/JFrog+Container+Registry+REST+API#JFrogContainerRegistryRESTAPI-ListDockerTags) rest APIs, set the artifactory.docker.catalogs.tags.fallback.fetch.remote.cache system property to true (default false) in the artifactory.system.properties file.
> 
> 
> 
> Have you got any hands-on experience with what this flag does (or doesn't do)?

In our experience this works when you define a single remote. When you create a virtual repo with multiple remotes it doesn't seem to send the request to any remotes. Reaching out to support validated this limitation.

I'll have a look at this property on our instance to validate.

**Comment by chezsmithy** [comment]
> > There is a further problem with jfrog. If you are using a virtual docker repo, the tags api calls will only return what's been pulled vs what is available upstream. Makes renovate fail to find new versions or digests.
> 
> 
> 
> @chezsmithy - I got some good news from JFrog Support about your comment.  They have already fixed this issue in https://www.jfrog.com/jira/browse/RTFACT-27278, with a public release slated for Q4-2022.  In the interim, they suggested to use the remote docker repository directly, instead of a virtual docker repository. That solution isn't too practical, so we're holding off until their public release of the fix.

Thanks for this comment. Replied a bit too soon. 

I looked at this fix you linked. It's not quite the same issue, but maybe they solved the problem. Renovate asks for all the tags upstream and all the digests. It expects the full list back. Jfrog with a virtual remote isn't passing anything upstream. I hope they have also fixed that. In the mean time we are looking for alternatives.

**Comment by setchy** [comment]
> I looked at this fix you linked. It's not quite the same issue, but maybe they solved the problem. 

I had the same initial thought, but double and triple checked with JFrog Support.  They've assured me that although that ticket isn't quite the same, that the root should be resolved and deployed shortly

**Review by rarkins — APPROVED** [review]

**Pull request merged**